### PR TITLE
Use PropertyName value for enum serialization/deserialization

### PR DESCRIPTION
**Pull request by wikes82**
Pull request for this issue : https://github.com/firebase/firebase-android-sdk/issues/239

### CODE REVIEW

**Comment by wikes82** [comment]
I signed CLA

**Comment by wikes82** [comment]
CLA signed

**Comment by wikes82** [comment]
please check CLA again

File: firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java [hunk 1]
```diff
@@ -324,6 +334,19 @@ public static Object convertToPlainJavaTypes(Object object) {
       String value = (String) object;
       // We cast to Class without generics here since we can't prove the bound
       // T extends Enum<T> statically
+
+      // try to use PropertyName if exist
+      Field[] enumFields = clazz.getFields();
+      for (Field field : enumFields) {
+        if (field.isAnnotationPresent(PropertyName.class)) {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/util/CustomClassMapper.java]: This should reuse `propertyName` (or at least `annotationName`).

Also, not all fields in an enum class are necessarily enum constants, so I think you need to check `Field.isEnumConstant`. That would make the inner part of this loop look like this:

```java
if (field.isEnumConstant()) {
  String propertyName = propertyName(field);
  if (value.equals(propertyName)) {
    value = field.getName();
    break;
  }
}
```

File: firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java [hunk 2]
```diff
@@ -709,7 +709,10 @@ public static void setValue1(String value1) {
 
   private enum ComplexEnum {
     One("one"),
-    Two("two");
+    Two("two"),
+
+    @PropertyName("three")
+    THREE("three");
```
Comment (wilhuff) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java]: Please make either the PropertyName or the value different here so there's no chance the test is accidentally succeeding.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java [hunk 3]
```diff
@@ -1974,18 +1979,22 @@ public void enumsAreSerialized() {
...
   public void enumsAreParsed() {
-    String json = "{'enumField': 'Bar', 'enumValue': 'Foo', 'complexEnum': 'One'}";
+    String json =
+        "{'enumField': 'Bar', 'enumValue': 'Foo', 'complexEnum': 'One', 'enumUsingPropertyName': 'three'}";
     EnumBean bean = deserialize(json, EnumBean.class);
     assertEquals(bean.enumField, SimpleEnum.Bar);
     assertEquals(bean.enumValue, SimpleEnum.Foo);
     assertEquals(bean.complexEnum, ComplexEnum.One);
+    assertEquals(bean.enumUsingPropertyName, ComplexEnum.THREE);
   }
```
Comment (wilhuff) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java]: Please add a test for a pathological case like this enum:

```java
enum PathologicalEnum {
  @PropertyName("Two")
  One,

  @PropertyName("One")
  Two,
}
```

**Comment by wikes82** [comment]
not sure why those 2 tests failed

**Comment by wilhuff** [comment]
Those aren't your fault. We've recently been having trouble with some of the tasks in those tests which has made them flaky. This submission wouldn't depend on getting those resolved.

File: firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java (hunk 2 above)
Comment (wikes82) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java]: I made this change.. PropertyName is Three and value is three

File: firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java (hunk 3 above)
Comment (wikes82) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/util/MapperTest.java]: added this test

**Pull request merged**